internal/discovery/discovery.go

```diff
@@ -0,0 +1,283 @@
+				}
+				r.RemoveFromMap(gotGVKP)
+				r.SafeWrite(func() {
+					r.WasUpdated = true
+				})
+				klog.InfoS("removed CRD for GVK", "group", g, "version", v, "kind", k)
+			}
+			r.SafeWrite(func() {
+				r.CRDsDeleteEvents++
+				r.CRDsCacheCount--
+			})
+			klog.InfoS("removed CRD", "total delete events", r.CRDsDeleteEvents, "total number of CRDs affecting cache", r.CRDsCacheCount)
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Respect context cancellation.
+	go func() {
+		for range ctx.Done() {
+			klog.InfoS("context cancelled, stopping discovery")
+			close(stopper)
+			return
+		}
+	}()
+	go informer.Run(stopper)
+	return nil
+}
+
+// ResolveGVKToGVKPs resolves the variable VKs to a GVK list, based on the current cache.
+func (r *CRDiscoverer) ResolveGVKToGVKPs(gvk schema.GroupVersionKind) (resolvedGVKPs []groupVersionKindPlural, err error) { // nolint:revive
+	g := gvk.Group
+	v := gvk.Version
+	k := gvk.Kind
+	if g == "" || g == "*" {
+		return nil, fmt.Errorf("group is required in the defined GVK %v", gvk)
+	}
+	hasVersion := v != "" && v != "*"
+	hasKind := k != "" && k != "*"
+	// No need to resolve, return.
+	if hasVersion && hasKind {
+		var p string
+		for _, el := range r.Map[g][v] {
+			if el.Kind == k {
+				p = el.Plural
+				break
+			}
+		}
+		return []groupVersionKindPlural{
+			{
+				GroupVersionKind: schema.GroupVersionKind{
+					Group:   g,
+					Version: v,
+					Kind:    k,
+				},
+				Plural: p,
+			},
+		}, nil
+	}
+	if hasVersion && !hasKind {
+		kinds := r.Map[g][v]
+		for _, el := range kinds {
+			resolvedGVKPs = append(resolvedGVKPs, groupVersionKindPlural{
+				GroupVersionKind: schema.GroupVersionKind{
+					Group:   g,
+					Version: v,
+					Kind:    el.Kind,
+				},
+				Plural: el.Plural,
+			})
+		}
+	}
+	if !hasVersion && hasKind {
+		versions := r.Map[g]
+		for version, kinds := range versions {
+			for _, el := range kinds {
+				if el.Kind == k {
+					resolvedGVKPs = append(resolvedGVKPs, groupVersionKindPlural{
+						GroupVersionKind: schema.GroupVersionKind{
+							Group:   g,
+							Version: version,
+							Kind:    k,
+						},
+						Plural: el.Plural,
+					})
+				}
+			}
+		}
+	}
+	if !hasVersion && !hasKind {
+		versions := r.Map[g]
+		for version, kinds := range versions {
+			for _, el := range kinds {
+				resolvedGVKPs = append(resolvedGVKPs, groupVersionKindPlural{
+					GroupVersionKind: schema.GroupVersionKind{
+						Group:   g,
+						Version: version,
+						Kind:    el.Kind,
+					},
+					Plural: el.Plural,
+				})
+			}
+		}
+	}
+	return
+}
+
+// PollForCacheUpdates polls the cache for updates and updates the stores accordingly.
+func (r *CRDiscoverer) PollForCacheUpdates(
+	ctx context.Context,
+	opts *options.Options,
+	storeBuilder *store.Builder,
+	m *metricshandler.MetricsHandler,
+	factoryGenerator func() ([]customresource.RegistryFactory, error),
+) {
+	// The interval at which we will check the cache for updates.
+	t := time.NewTicker(Interval)
```

Comment:
I would have the ticker as a property on the `CRDiscoverer`, so that we can inject a fake one for tests.

Reply:
`PollForCacheUpdates` doesn't return anything, and this is the only place where we use the interval ticker. I'm not sure what are we going to test for here.

Reply:
(assuming this is about unit tests, otherwise we'd have to pass our own `CRDiscoverer` down `RunKubeStateMetricsWrapper` IIUC, since that's where the control flow starts in the e2e suite currently (setting up KSM, etc.))

Reply:
ACK, I'll push this in a follow up PR. 👍🏼